cmd/kubectl-gadget/deploy.go

```diff
@@ -362,35 +192,105 @@ func runDeploy(cmd *cobra.Command, args []string) error {
-		fallbackPodInformer,
+	traceObjects, err := parseK8sYaml(resources.TracesCustomResource)
+	if err != nil {
+		return err
 	}
 
-	var w io.Writer = os.Stdout
+	objects = append(objects, traceObjects...)
 
-	if file != "" {
-		f, err := os.OpenFile(file, os.O_CREATE|os.O_WRONLY, 0o644)
-		if err != nil {
-			return fmt.Errorf("failed to create file: %w", err)
-		}
-		defer f.Close()
-		w = f
+	config, err := utils.KubernetesConfigFlags.ToRESTConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create RESTConfig: %w", err)
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return err
 	}
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(discoveryClient))
 
-	fmt.Fprintf(w, "%s\n---\n", resources.TracesCustomResource)
-	err = t.Execute(w, p)
+	dynamicClient, err := dynamic.NewForConfig(config)
 	if err != nil {
-		return fmt.Errorf("failed to generate deploy template: %w", err)
+		return fmt.Errorf("failed to create dynamic client: %w", err)
+	}
+
+	for _, object := range objects {
+		switch object.(type) {
+		case *v1.Namespace:
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		case *v1.ServiceAccount:
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		case *rbacv1.Role:
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		case *rbacv1.RoleBinding:
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		case *rbacv1.ClusterRole:
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		case *rbacv1.ClusterRoleBinding:
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		case *apiextv1.CustomResourceDefinition:
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		case *appsv1.DaemonSet:
+			daemonSet := object.(*appsv1.DaemonSet)
+			gadgetContainer := &daemonSet.Spec.Template.Spec.Containers[0]
+
+			gadgetContainer.Image = image
+
+			policy, err := stringToPullPolicy(imagePullPolicy)
+			if err != nil {
+				return err
+			}
+			gadgetContainer.ImagePullPolicy = policy
+
+			if !livenessProbe {
+				gadgetContainer.LivenessProbe = nil
+			}
+
+			for i, _ := range gadgetContainer.Env {
+				switch gadgetContainer.Env[i].Name {
+				case "GADGET_IMAGE":
+					gadgetContainer.Env[i].Value = image
+				case "INSPEKTOR_GADGET_VERSION":
+					gadgetContainer.Env[i].Value = version
+				case "INSPEKTOR_GADGET_OPTION_HOOK_MODE":
+					gadgetContainer.Env[i].Value = hookMode
+				case "INSPEKTOR_GADGET_OPTION_FALLBACK_POD_INFORMER":
+					gadgetContainer.Env[i].Value = strconv.FormatBool(fallbackPodInformer)
+				}
+			}
+
+			err = createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
What about something like: 

```suggestion
	for _, object := range objects {
		if daemonSet, ok := object.(*appsv1.DaemonSet); ok {
			gadgetContainer := &daemonSet.Spec.Template.Spec.Containers[0]

			gadgetContainer.Image = image

			policy, err := stringToPullPolicy(imagePullPolicy)
			if err != nil {
				return err
			}
			gadgetContainer.ImagePullPolicy = policy

			if !livenessProbe {
				gadgetContainer.LivenessProbe = nil
			}

			for i, _ := range gadgetContainer.Env {
				switch gadgetContainer.Env[i].Name {
				case "GADGET_IMAGE":
					gadgetContainer.Env[i].Value = image
				case "INSPEKTOR_GADGET_VERSION":
					gadgetContainer.Env[i].Value = version
				case "INSPEKTOR_GADGET_OPTION_HOOK_MODE":
					gadgetContainer.Env[i].Value = hookMode
				case "INSPEKTOR_GADGET_OPTION_FALLBACK_POD_INFORMER":
					gadgetContainer.Env[i].Value = strconv.FormatBool(fallbackPodInformer)
				}
			}
		}

		err = createResource(dynamicClient, mapper, object)
		if err != nil {
			return err
		}
	}
```